xdsl/transforms/canonicalization_patterns/arith.py

```diff
@@ -16,3 +17,28 @@ def match_and_rewrite(self, op: arith.Addi, rewriter: PatternRewriter) -> None:
             and value.value.data == 0
         ):
             rewriter.replace_matched_op([], [op.rhs])
+
+
+class FoldConstConstOp(RewritePattern):
+    @op_type_rewrite_pattern
+    def match_and_rewrite(
+        self, op: arith.FloatingPointLikeBinaryOperation, rewriter: PatternRewriter, /
+    ):
+        if (
+            isinstance(op.lhs.owner, arith.Constant)
+            and isinstance(op.rhs.owner, arith.Constant)
+            and isa(l := op.lhs.owner.value, builtin.AnyFloatAttr)
+            and isa(r := op.rhs.owner.value, builtin.AnyFloatAttr)
+        ):
+            match type(op):
+                case arith.Addf:
+                    val = l.value.data + r.value.data
+                case arith.Subf:
+                    val = l.value.data - r.value.data
+                case arith.Mulf:
+                    val = l.value.data * r.value.data
+                case arith.Divf:
+                    val = l.value.data / r.value.data
```

Comment:
What happens if `val` is `NaN`?
We should probably check for it and only do the replacement if `val` is not `NaN`, as `arith.constant` doesn't support `NaN` afaik?

Reply:
Excellent point. Theoretically, python should handle it. For instance, `3 / float('nan')` will evaluate to `nan`.

However, I can't seem to get nan to round-trip to verify this. `str(builtin.FloatAttr(float('nan'), 32))` returns `'nan : f32'`, but `arith.constant nan : f32` will throw a parser error saying:
```
    %a = arith.constant nan : f32
                        ^^^
                        attribute expected
```
Am I doing this right?

Reply:
can mlir parse that? it's possible that xDSL has a bug here

Reply:
Do you have the bandwidth to fix xDSL printing of nans in that case?

Reply:
It's not straightforward, xdsl likely parses hexadecimal float literals as hexadecimal int literals, from which it builds a float rather than using it as a bit-string. See the [floatattr](https://mlir.llvm.org/docs/Dialects/Builtin/#floatattr) docs